test/persistent_receiver_test.go

```diff
@@ -1696,19 +1694,46 @@ var _ = Describe("PersistentReceiver", func() {
 
 				It("handles unsolicited terminations while already terminating", func() {
 					blocker := make(chan struct{})
+                    signal := make(chan bool)
 					messageChannel := make(chan message.InboundMessage, numMessages+1)
+                    waitToReceiveMessages := 30 * time.Second // This timer needs to be big enough to give us a very
+                                                              // high chance of both receiving the messages and getting
+                                                              // past the state change in Terminate. It may need to be
+                                                              // updated in the future since this only mitigates the
+                                                              // race condition and doesn't eliminate it.
 					messageReceiver.ReceiveAsync(func(inboundMessage message.InboundMessage) {
+                        loop:
+                            for timeout := time.After(waitToReceiveMessages); ; {
+                                select {
+                                case <-timeout:
+                                    break loop
+                                default:
+                                }
+                                if !messageReceiver.IsRunning() {
+                                    break loop
+                                }
+                            }
+                        signal <- true
 						<-blocker
 						messageChannel <- inboundMessage
 					})
 					helpers.PublishNPersistentMessages(messagingService, topicString, numMessages+1)
 					helpers.ValidateMetric(messagingService, metrics.PersistentMessagesReceived, numMessages+1)
-					terminateDuration := 2 * time.Second
+                    terminateDuration := 2 * time.Second // This timer does not need to be especially long, since we're
+                                                         // only needed a fraction of a second to block on signal
+                                                         // and wait for that channel to be written to in the receiver
+                                                         // callback.
 					terminateChan := messageReceiver.TerminateAsync(terminateDuration)
+                    // Blocking on this signal prevents us from proceeding with the test unless either the timer has
+                    // expired or the state transition has occured.
+                    <-signal
+                    Expect(messageReceiver.IsRunning()).To(BeFalse())
+                    //Eventually(messageReceiver.IsRunning(), 5 * time.Second).Should(BeFalse())
 					terminationFunction(messagingService, messageReceiver)
 					Consistently(terminationReceived).ShouldNot(Receive())
 					time.Sleep(terminateDuration)
 					close(blocker)
+                    close(signal)
```

Comment:
I don't think this `signal` channel is useful.

Reply:
The problem with the test before these changes is that there was no mechanism guaranteeing that the receiver was actually terminating or terminated when the queue was deleted and unsolicited termination was triggered. This caused a race condition where the test actually needed to work incorrectly in order to pass, since the queue name would be retrieved from the still running receiver before Terminate() had even been called from the TerminateAsync goroutine. We happened to get a single correct run of this test in automation where Terminate() actually ran in the separate goroutine before the queue name was accessed in termination_function(). To solve the failure, all I had to do was change where the termination_function() was getting the queue info from. After that, I changed the test to both pass AND test unsolicited termination during solicited termination.

The signal channel is essential to make this test work correctly, because we need to block the main thread and not call termination_function until we are certain the Termintate() has been called. The only point that we can block in Terminate() in an integration test is the receiver callback. We cannot do this through a waiter/poll on receiver state on the main thread, because it is possible for the main thread to spin and exit the timeout without Terminate() ever actually being called on the separate goroutine. The only way in GoLang to ensure that the current thread does not execute until a condition is met is to block the main goroutine on a channel read, and then have a separate goroutine execute the timed state poll and write to the channel that blocks the main goroutine, unblocking the main goroutine. The 'blocker' channel ensures that the call to Terminate() does not complete unless either the grace period passed to TerminateAsync() expires or the test completes and the 'blocker' channel is closed. Therefore, we know that Terminate() has started, but not yet completed, when we delete the queue and trigger unsolicited termination, which is what the test was supposed to be doing in the first place. Do you have any other questions about the usefulness of the signal channel?

Reply:
This is fine then.